joern-cli/frontends/c2cpg/src/main/scala/io/joern/c2cpg/passes/HeaderContentPass.scala

```diff
@@ -0,0 +1,117 @@
+import overflowdb.traversal._
+
+class HeaderContentPass(cpg: Cpg, keyPool: Option[KeyPool], config: Config) extends CpgPass(cpg, keyPool = keyPool) {
+
+  private val systemIncludePaths = IncludeAutoDiscovery.discoverIncludePaths(config)
+
+  private val absolutePath: String = File(config.inputPaths.head).path.toAbsolutePath.normalize().toString
+  private val filename: String = s"$absolutePath:<includes>"
+  private val globalName: String = NamespaceTraversal.globalNamespaceName
+  private val fullName: String = MetaDataPass.getGlobalNamespaceBlockFullName(Some(filename))
+
+  private def setExternal(node: HasFilename, diffGraph: DiffGraph.Builder): Unit = {
+    if (node.isInstanceOf[HasIsExternal] && systemIncludePaths.exists(p => node.filename.startsWith(p.toString))) {
+      diffGraph.addNodeProperty(node.asInstanceOf[StoredNode], PropertyNames.IS_EXTERNAL, Boolean.box(true))
+    }
+  }
+
+  private def createGlobalBlock(dstGraph: DiffGraph.Builder): NewBlock = {
+    val includesFile = NewFile().name(filename).order(0)
+
+    val namespaceBlock = NewNamespaceBlock()
+      .name(globalName)
+      .fullName(fullName)
+      .filename(filename)
+      .order(1)
+
+    val fakeGlobalIncludesMethod =
+      NewMethod()
+        .name(globalName)
+        .code(globalName)
+        .fullName(fullName)
+        .filename(filename)
+        .lineNumber(1)
+        .astParentType(NodeTypes.NAMESPACE_BLOCK)
+        .astParentFullName(fullName)
+
+    val blockNode = NewBlock()
+      .order(1)
+      .argumentIndex(1)
+      .typeFullName("ANY")
+
+    val methodReturn = NewMethodReturn()
+      .code("RET")
+      .evaluationStrategy(EvaluationStrategies.BY_VALUE)
+      .typeFullName("ANY")
+      .order(2)
+
+    val ast = Ast(includesFile).withChild(
+      Ast(namespaceBlock).withChild(
+        Ast(fakeGlobalIncludesMethod).withChild(Ast(blockNode)).withChild(Ast(methodReturn))
+      )
+    )
+
+    Ast.storeInDiffGraph(ast, dstGraph)
+    blockNode
+  }
+
+  private def createMissingAstEdges(dstGraph: DiffGraph.Builder): Unit = {
+    val globalBlock = createGlobalBlock(dstGraph)
+    Traversal(cpg.graph.nodes()).whereNot(_.inE(EdgeTypes.AST)).foreach {
+      case srcNode: HasFilename if FileDefaults.isHeaderFile(srcNode.filename) =>
+        dstGraph.addEdgeToOriginal(globalBlock, srcNode.asInstanceOf[StoredNode], EdgeTypes.AST)
+        setExternal(srcNode, dstGraph)
+      case srcNode: Local =>
+        dstGraph.addEdgeToOriginal(globalBlock, srcNode.asInstanceOf[StoredNode], EdgeTypes.AST)
+      case _ =>
+    }
+  }
+
+  private def typeNeedsTypeDeclStub(t: Type): Boolean =
+    cpg.graph.nodes(NodeTypes.TYPE_DECL).has(Properties.FULL_NAME, t.typeDeclFullName).isEmpty
+
+  private def createMissingTypeDecls(dstGraph: DiffGraph.Builder): Unit = {
+    Traversal(cpg.graph.nodes(NodeTypes.TYPE))
+      .foreach {
+        case t: Type if typeNeedsTypeDeclStub(t) =>
+          val newTypeDecl = NewTypeDecl()
```

Comment:
```suggestion
    Traversal(cpg.graph.nodes(NodeTypes.TYPE))
      .filter(typeNeedsTypeDeclStub)
      .foreach { t =>
        val newTypeDecl = NewTypeDecl()
```
matter of personal style preference, so feel free to ignore, but IMO the above is more readable and concise. you can leave out the `case _` as well.

Reply:
Hm... we need to know that `t` is a `Type` to have access to `name` and `typeDeclFullName`.
Hence, the explicit `case`.

Reply:
ah i see, yeah that makes sense, ignore my comment :)